{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5d893234",
   "metadata": {},
   "source": [
    "# 1. Build Kallisto Index Using kb-ref
Use the primary DNA assembly and GTF file as inputs.
2. Install kb Python Package
Ensure the kb Python package is installed.
3. Run Command in Shell Script or Terminal
Execute the command in a shell script or directly in the terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bcda945",
   "metadata": {},
   "source": [
    
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83bc98b6",
   "metadata": {},
   "source": [
   
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73842d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "kb ref -i ref_index.idx -g t2g.txt -f1 cdna.fa -f2 intron.fa -c1 spliced.txt -c2 unspliced.txt  --workflow=lamanno /oak/stanford/groups/icobos/Akila/ref_index/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz /oak/stanford/groups/icobos/Akila/ref_index/Homo_sapiens.GRCh38.105.gtf.gz \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04d2e573",
   "metadata": {},
   "source": [
    "# Count the abundance using kb count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cb5f9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load /oak/stanford/groups/icobos/miniconda3/bin/kallisto\n",
    "module load /oak/stanford/groups/icobos/miniconda3/bin/kb\n",
    "### specify input sample directory folder name\n",
    "declare -a StringArray=(\"input samples\")\n",
    "\n",
    "for val in ${StringArray[@]}; do\n",
    "    # create directory for each samples\n",
    "    mkdir=/oak/stanford/groups/icobos/Akila/count/$val\n",
    "    cd  /oak/stanford/groups/icobos/RawSequencingDataADnuclei/$val\n",
    "    # count the abundance for each sample and output the counts in directory\n",
    "    kb count --h5ad -i /oak/stanford/groups/icobos/Akila/refbuild_105/ref_index.idx -g /oak/stanford/groups/icobos/Akila/refbuild_105/t2g.txt -x 10xv2 -o /oak/stanford/groups/icobos/Akila/count/$val -c1 /oak/stanford/groups/icobos/Akila/refbuild_105/spliced.txt -c2 /oak/stanford/groups/icobos/Akila/refbuild_105/unspliced.txt --workflow=lamanno --filter bustools $(find -type f -name '*.fastq.gz'| sort)\n",
    "    done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3686317",
   "metadata": {},
   "source": [
    "# Add metadata to the anndata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47d910f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata \n",
    "from scipy.io import mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "# list the filenames\n",
    "\n",
    "list1=[\"input samples\"]\n",
    "\n",
    "# output file names\n",
    "list2=[\"output filenames\"]\n",
    "\n",
    "#read metadata from csv files\n",
    "metadata=pd.read_csv(\"/oak/stanford/groups/icobos/Akila/sample_metadata.csv\")\n",
    "\n",
    "#Read all adata and add metadata information for all h5ad file\n",
    "\n",
    "for i in range(len(list1)):\n",
    "\n",
    "    adata = anndata.read_h5ad(\"/oak/stanford/groups/icobos/Akila/count/\"+str(list1[i])+\"/counts_unfiltered/adata.h5ad\")\n",
    "    adata.var_names_make_unique()\n",
    "    S1=metadata.loc[i]\n",
    "    \n",
    "  # write metadata \n",
    "    list1[i]=adata\n",
    "    list1[i].obs['Unique.identifier']=S1[0]\n",
    "    list1[i].obs['Assay']=S1[1]\n",
    "    list1[i].obs['Sequencing.platform']=S1[2]\n",
    "    list1[i].obs['Read2.length']=S1[3]\n",
    "    list1[i].obs['Tissue.processing.date']=S1[4]\n",
    "    list1[i].obs['Donor']=S1[5]\n",
    "    list1[i].obs['subject']=S1[6]\n",
    "    list1[i].obs['APOE']=S1[7]\n",
    "    list1[i].obs['Brain.Region']=S1[8]\n",
    "    list1[i].obs['SORT']=S1[9]\n",
    "    list1[i].obs['NP.Diagnosis']=S1[10]\n",
    "    list1[i].obs['Braak.stage']=S1[11]\n",
    "    list1[i].obs['Disease.Group']=S1[12]\n",
    "    list1[i].obs['Amyloid']=S1[13]\n",
    "    list1[i].obs['Brain.weight']=S1[14]\n",
    "    list1[i].obs['PMI.hr.']=S1[15]\n",
    "    list1[i].obs['Race']=S1[16]\n",
    "    list1[i].obs['Age']=S1[17]\n",
    "    list1[i].obs['Sex']=S1[18]\n",
    "    list1[i].obs['RIN']=S1[19]\n",
    "    \n",
    "    \n",
    "    \n",
    " #write the updated obs and h5ad file   \n",
    "\n",
    "    list1[i].write(\"/scratch/groups/icobos/Akila/metadata/\"+str(list2[i])+\".h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f402b3ab",
   "metadata": {},
   "source": [
    "# Convert anndata counts as matrix format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b2fadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.sparse as sparse\n",
    "import scipy.io as sio\n",
    "import scipy.stats as stats\n",
    "import numpy as np\n",
    "import anndata\n",
    "\n",
    "list1=[\"input sample files\"]\n",
    "\n",
    "\n",
    "for i in range(len(list1)):\n",
    "    #load h5ad file\n",
    "    adata = anndata.read_h5ad(\"/scratch/groups/icobos/Akila/metadata/\"+str(list1[i])+\".h5ad\")\n",
    "    # convert sparse matrix to mtx format\n",
    "    adata.X=adata.layers[\"spliced\"]+adata.layers[\"unspliced\"]\n",
    "    sio.mmwrite(\"/scratch/groups/icobos/Akila/Matrix/\"+str(list1[i])+\".mtx\",data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4ff4bcc",
   "metadata": {},
   "source": [
    "# Detect empty droplets -R package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bc27016",
   "metadata": {},
   "outputs": [],
   "source": [
    "## load R Packages\n",
    "library(DropletUtils,lib.loc=\"/oak/stanford/groups/icobos/Akila/Rpackages/\")\n",
    "library(Matrix)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(ggExtra)\n",
    "\n",
    "list1<-c(\"input sample file names\")\n",
    "\n",
    "list_count <- length(list1)\n",
    "\n",
    "print(list_count)\n",
    "\n",
    "for (i in list_count)\n",
    "{\n",
    "    ## load matrix for each sample\n",
    "    \n",
    "  m <- readMM(paste0(\"/scratch/groups/icobos/Akila/Astrocytes/Matrix/\",list1[i],\".mtx\"))\n",
    "  m <- Matrix::t(m)\n",
    "  m <- as(m, \"dgCMatrix\")\n",
    "    \n",
    "    ## calculate empty droplets based on barcodes \n",
    "    \n",
    "  tot_counts <- Matrix::colSums(m)  \n",
    "  bc_rank <- barcodeRanks(m)\n",
    "  set.seed(200)\n",
    "  e.out <- emptyDrops(m)\n",
    "    \n",
    "    ## test each cell for real/ambient RNA profile\n",
    "  is.cell <- e.out$FDR <= 0.05\n",
    "  sum(is.cell, na.rm=TRUE)\n",
    "  #write e.out\n",
    "    \n",
    "  write.csv(bc_rank,paste(\"/scratch/groups/icobos/Akila/Matrix/rank/\",list1[i],\".csv\"))\n",
    "\n",
    "  write.csv(e.out,paste(\"/scratch/groups/icobos/Akila/Matrix/e_out/\",list1[i],\".csv\"))\n",
    "  mypath <- file.path(\"/scratch/groups/icobos/Akila/Matrix/logprobability/\",paste(list1[i], \".jpg\", sep = \"\"))\n",
    "    \n",
    "  #log probaility plot empty droplet testing\n",
    "  jpeg(file=mypath)\n",
    "  par(mar=c(5,4,1,1), mfrow=c(1,2), bty=\"n\")\n",
    "    plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, \"red\", \"black\"),\n",
    "    xlab=\"Total UMI count\", ylab=\"-Log Probability\", cex=0.2)\n",
    "    abline(v = bc_rank@metadata$inflection, col=\"darkgreen\")\n",
    "    abline(v = bc_rank@metadata$knee, col=\"dodgerblue\")\n",
    "    legend(\"bottomright\", legend=c(\"Inflection\", \"Knee\"), bty=\"n\", \n",
    "    col=c(\"darkgreen\", \"dodgerblue\"), lty=1, cex=1.2)\n",
    "    \n",
    "    #zoom in low UMI part\n",
    "    #log probaility plot empty droplet testing\n",
    "  \n",
    "    plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, \"red\", \"black\"),\n",
    "    xlab=\"Total UMI count\", ylab=\"-Log Probability\", cex=0.2, xlim=c(0,2200), ylim=c(0,2200))\n",
    "    abline(v = bc_rank@metadata$inflection, col=\"darkgreen\")\n",
    "    abline(v = bc_rank@metadata$knee, col=\"dodgerblue\")\n",
    "    dev.off()\n",
    "    \n",
    "    #cell_calling plots-empty droplets    \n",
    "    set.seed(100)\n",
    "    #set inflection point\n",
    "    limit <- bc_rank@metadata$inflection  \n",
    "    all.out <- emptyDrops(m, lower=limit, test.ambient=TRUE)\n",
    "    \n",
    "    #write all.out\n",
    "    write.csv(all.out,paste(\"/scratch/groups/icobos/Akila/Matrix/all_out/\",list1[i],\".csv\"))\n",
    "    mypath1 <- file.path(\"/scratch/groups/icobos/Akila/Matrix/cell_calling/\",paste(list1[i], \".jpg\", sep = \"\"))\n",
    "    jpeg(file=mypath1)\n",
    "    \n",
    "    #histogram null hypotheis -low UMI denotes empty droplets!!!!\n",
    "    hist(all.out$PValue[all.out$Total <= limit & all.out$Total > 0],\n",
    "    xlab=\"P-value\", main=\"\", col=\"grey80\")\n",
    "    dev.off()\n",
    "    \n",
    "    # cut off write\n",
    "    both = which(is.cell & e.out$Total >= bc_rank@metadata$inflection)\n",
    "    filt = m[,both]\n",
    "    \n",
    "    #using inflection\n",
    "    inflect = which(bc_rank$total >= bc_rank@metadata$inflection)\n",
    "    inflect = m[,inflect]\n",
    "    \n",
    "    #using(knee)\n",
    "    kn = which(bc_rank$total >= bc_rank@metadata$knee)\n",
    "    kn = m[,kn]\n",
    "    print(paste(list1[i],\"FDR<0.05\",nrow(is.cell),ncol(is.cell)))\n",
    "    #print(paste(list1[i],\"FDR<0.05+inflection\",nrow(filt)))\n",
    "    print(paste(list1[i],\"Inflection\",nrow(inflect),ncol(inflect)))\n",
    "    print(paste(list1[i],\"Knee\",nrow(kn),ncol(kn)))\n",
    "    print(paste(list1[i],\"total cell counts\",nrow(bc_rank),ncol(bc_rank)))\n",
    "  }\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01610769",
   "metadata": {},
   "source": [
    "# Add empty droplet details in anndata object & Compute Mitochondrial and ribosomal gene frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e42b231",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata \n",
    "from scipy.io import mmread\n",
    "import scrublet as scr\n",
    "import doubletdetection\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import sparse, io\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "\n",
    "list1=[\"input file names\"]\n",
    "\n",
    "for i in range(len(list1)):\n",
    "    \n",
    "    ##read h5ad file\n",
    "\n",
    "    adata = anndata.read_h5ad(\"/scratch/groups/icobos/Akila/counts/metadata/\"+str(list1[i])+\".h5ad\")\n",
    "   # bdata=anndata.read_h5ad(\"/scratch/groups/icobos/Akila/unsplice_counts/\"+str(list1[i])+\".h5ad\")\n",
    "    \n",
    "    ## read bc_rank,FDR from R emptydroplet\n",
    "    \n",
    "    bc_rank=pd.read_csv(\"/scratch/groups/icobos/Akila/Matrix/rank/\"+\" \"+str(list1[i])+\" \"+\".csv\")\n",
    "    FDR=pd.read_csv(\"/scratch/groups/icobos/Akila/Matrix/e_out/\"+\" \"+str(list1[i])+\".csv\")\n",
    "    \n",
    "    ## copy information (FDR,rank)inside h5ad obs\n",
    "    \n",
    "    adata.obs.index = bc_rank.index\n",
    "    adata.obs.index = FDR.index\n",
    "    adata.obs['rank'] = bc_rank['rank']\n",
    "    adata.obs[\"droplet_FDR\"]=FDR[\"FDR\"]\n",
    "    adata.obs[\"total_counts\"]=bc_rank[\"total\"]\n",
    "    \n",
    "    ## write the metadata along with h5ad\n",
    "    adata.write(\"/scratch/groups/icobos/Akila/Astrocytes/Matrix/\"+str(list1[i])+\".h5ad\")\n",
    "    \n",
    "       \n",
    "    \n",
    "    ### read the h5ad file\n",
    "    adata = anndata.read_h5ad(\"/scratch/groups/icobos/Akila/Astrocytes/Matrix/\"+str(list1[i])+\".h5ad\")\n",
    "    \n",
    "   \n",
    "    ## remove cells based on FDR <0.05 (remove emptydroplets)\n",
    "    adata=adata[adata.obs['droplet_FDR'] < 0.05,:]\n",
    "    \n",
    "    ## merge the spliced and unspliced counts and store in nuclear layer\n",
    "    \n",
    "    adata.layers[\"nuclear\"]=adata.layers[\"spliced\"]+adata.layers[\"unspliced\"]\n",
    "    \n",
    "    ### Calculate QC metrices\n",
    "    adata.X = csr_matrix(adata.layers['nuclear'])\n",
    "    \n",
    "      #mitochondrial genes\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('MT-') \n",
    "    \n",
    "    # ribosomal genes\n",
    "    adata.var['ribo'] = adata.var_names.str.startswith((\"RPS\",\"RPL\"))\n",
    "    \n",
    "    # hemoglobin genes.\n",
    "    adata.var['hb'] = adata.var_names.str.contains((\"^HB[^(P)]\"))\n",
    "\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt','ribo','hb'],percent_top=None, log1p=False, inplace=True)\n",
    "    \n",
    "    mito_genes = adata.var_names.str.startswith('MT-')\n",
    "    \n",
    "    # for each cell compute fraction of counts in mito genes vs. all genes\n",
    "    # the `.A1` is only necessary as X is sparse (to transform to a dense array after summing)\n",
    "    adata.obs['pct_counts_mt'] = np.sum(\n",
    "        adata[:, mito_genes].X, axis=1).A1/np.sum(adata.X, axis=1).A1\n",
    "    \n",
    "    # add the total counts per cell as observations-annotation to adata\n",
    "    adata.obs['n_counts'] = adata.X.sum(axis=1).A1\n",
    "\n",
    "    ##  save the plots\n",
    "    sc.pl.violin(adata, ['n_genes_by_counts', 'n_counts', 'pct_counts_mt'],jitter=0.4, multi_panel=True)\n",
    "    plt.savefig(\"/scratch/groups/icobos/Akila/QC_plots/\"+str(list1[i])+\".png\")\n",
    "    \n",
    "    ## count the min_max_gene counts/total UMI counts for each sample(count number of cell after QC)\n",
    "    ngenes=adata.obs[\"n_genes_by_counts\"]\n",
    "    med=np.median(ngenes)\n",
    "    total=adata.obs[\"total_counts\"]\n",
    "    med1=np.median(total)\n",
    "    print(list1[i],min(ngenes),max(ngenes),round(sum(ngenes)/len(ngenes)),med,min(total),max(total),round(sum(total)/len(total)),med1,adata.shape)\n",
    "\n",
    "    #adata.write(\"/scratch/groups/icobos/Akila/QC_splice/\"+str(list1[i])+\".h5ad\")\n",
    "    #adata = anndata.read_h5ad(\"/scratch/groups/icobos/Akila/QC_splice/\"+str(list1[i])+\".h5ad\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2b6a4b",
   "metadata": {},
   "source": [
    "# Detect doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0207c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#predict doublets add the metadata\n",
    "    \n",
    "    clf = doubletdetection.BoostClassifier(\n",
    "    n_iters=10,\n",
    "    clustering_algorithm=\"louvain\",\n",
    "    standard_scaling=True,\n",
    "    pseudocount=0.1,\n",
    "    n_jobs=-1,)\n",
    "    doublets = clf.fit(adata.X).predict(p_thresh=1e-16, voter_thresh=0.5)\n",
    "    doublet_score = clf.doublet_score()\n",
    "    adata.obs[\"doublefinder\"] = doublets\n",
    "    adata.obs[\"doubletfinder_score\"] = doublet_score\n",
    "    ### write output\n",
    "    adata.write(\"/scratch/groups/icobos/Akila/doublets/\"+str(list1[i])+\".h5ad\")  \n",
    "    \n",
    "    #convergence plot\n",
    "    f = doubletdetection.plot.convergence(clf, save=\"/scratch/groups/icobos/Akila/QC_plots/doublets/convergence/\"+str(list1[i])+\".pdf\", show=True, p_thresh=1e-16, voter_thresh=0.5)\n",
    "    f3 = doubletdetection.plot.threshold(clf, save=\"/scratch/groups/icobos/Akila/QC_plots/doublets/threshold/\"+str(list1[i])+\".pdf\", show=True, p_step=6)\n",
    "   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f66f68c",
   "metadata": {},
   "source": [
    "# Concatenate all the anndata objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "794e29a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata \n",
    "from scipy.io import mmread\n",
    "\n",
    "###filenames=[\"Sample_100\",\"Sample_103\",\"Sample_110\",\"Sample_114\",\"Sample_124\",\"Sample_14\",\"Sample_18\",\"Sample_33\",\"Sample_46\",\"Sample_68\",\"Sample_92\",\"Sample_108\",\"Sample_112\",\"Sample_116\",\"Sample_138\",\"Sample_140\",\"Sample_30\",\"Sample_35\",\"Sample_59\",\"Sample_74\",\"Sample_94\"]\n",
    "filenames=\n",
    "adatas = [anndata.read_h5ad(\"/scratch/groups/icobos/Akila/Astrocytes/QC_plots/\"+filename+\".h5ad\") for filename in filenames]\n",
    "   # adatas = [sc.read_10x_h5(filename, gex_only = True) for filename in filenames]\n",
    "\n",
    "adata = adatas[0].concatenate(adatas[1:], batch_key='filename', batch_categories=filenames)\n",
    "\n",
    "adata.write(\"/oak/stanford/groups/icobos/Akila/Astrocytes/After_QC_nuclei_combined.h5ad\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c37ed53",
   "metadata": {},
   "source": [
    "# Harmony integration and clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c39c861",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy.external as sce\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from scipy.io import mmread\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "adata=anndata.read_h5ad(\"/oak/stanford/groups/icobos/Akila/Shell/scripts/After_QC_nuclei.h5ad\")\n",
    "\n",
    "#### copy the gene name instead of ENSG id (as index)\n",
    "adata.var[\"index1\"]=adata.var.index\n",
    "adata.var[\"gene_name\"]=adata.var[\"gene_name\"].astype('object')\n",
    "adata.var.loc[adata.var['gene_name'] ==\"\",'gene_name']=adata.var['index1']\n",
    "adata.var.set_index('gene_name',inplace=True)\n",
    "\n",
    "#### Filter minimum genes\n",
    "sc.pp.filter_cells(adata, min_genes=300)\n",
    "print(\"After removing min_genes\",adata.n_obs, adata.n_vars)\n",
    "\n",
    "\n",
    "#### scale Nrmalize data\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "### Find HVG\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "\n",
    "#### PCA \n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(adata, log=True)\n",
    "\n",
    "### Harmony integration mention covariates\n",
    "\n",
    "sce.pp.harmony_integrate(adata, ['Assay','Brain.Region'])\n",
    "\n",
    "#### check the Elbow plot for PCA\n",
    "### generate leiden graph based nearsest neighbor\n",
    "\n",
    "sc.pp.neighbors(adata, use_rep=\"X_pca_harmony\",n_pcs=35)\n",
    "sc.tl.leiden(adata,resolution = 1.0, key_added = \"leiden_1.0\")\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata,color=[\"Assay\", \"leiden_1.0\"],ncols=2, frameon=False,)\n",
    "plt.savefig(\"/scratch/groups/icobos/Akila/integration/harmony/new/min_300/After_Assay_brainregion.png\")\n",
    "\n",
    "### List the major celltype marker genes\n",
    "\n",
    "marker_genes_dict = {\n",
    "'Oligodendrocytes':['PLP1','MOG','ST18','MBP',\t'OLIG1',\t'COL18A1','OPALIN'],\n",
    "'OPC': ['PDGFRA','VCAN'],\n",
    "'Astrocytes':['SLC1A2','GFAP','CD44','CHI3L1','GRIA1']\n",
    "'Microglia':['CSF1R','CX3CR1','TREM2'],\n",
    "'Endothelial': \t['FLT1',\t'ERG'],\n",
    "'pericytes':\t['NOTCH3'],\n",
    "'VLMC':\t['CYP1B1',\t'COL1A2'],\n",
    "'Excitatory':\t['SLC17A7','CUX2','FEZF2','THEMIS','RORB','PCP4'],\n",
    "'Inhibitory':['GAD1','GAD2','ADARB2','LHX6','SST','PVALB','KIT']           \n",
    "             }\n",
    "\n",
    "\n",
    "### Annotate the major celltypes based on known markers\n",
    "sc.pl.dotplot(adata, marker_genes_dict,'leiden_1.0', dendrogram=True,standard_scale=\"var\")\n",
    "plt.savefig(\"/scratch/groups/icobos/Akila/integration/harmony/marker_harmony_leiden_1.0_dotplot.png\")\n",
    "\n",
    "### plot the Assay, total_counts\n",
    "sc.pl.umap(adata,color=[\"Assay\", \"leiden_1.0\",\"total_counts\"], frameon=False,)\n",
    "plt.savefig(\"/scratch/groups/icobos/Akila/integration/harmony/new/min_300/plot_new_leiden1.0_Assay_leiden.png\")\n",
    "\n",
    "#### calculate MArker\n",
    "adata.uns['log1p'][\"base\"] = None\n",
    "sc.tl.rank_genes_groups(adata,\"leiden_1.0\",method='wilcoxon',pts=True,corr_method='benjamini-hochberg')\n",
    "sc.tl.filter_rank_genes_groups(adata,min_in_group_fraction=0.2,min_fold_change=0.8)\n",
    "sc.tl.dendrogram(adata,\"leiden_1.0\")\n",
    "\n",
    "#### calculate marker and plot them\n",
    "sc.pl.rank_genes_groups_dotplot(adata,n_genes=25,key='rank_genes_groups_filtered',cmap='bwr',standard_scale='var')\n",
    "\n",
    "#### Annotate the anndata based on major celltype markers and include them in anndata. obs slot Major_celltypes\n",
    "#### save the marker genes\n",
    "result = adata.uns['rank_genes_groups_filtered']\n",
    "groups = result['names'].dtype.names\n",
    "ranked_genes_df = pd.DataFrame(\n",
    "    {group + '_' + key: result[key][group]\n",
    "     for group in groups for key in ['names', 'pvals_adj', 'logfoldchanges']}\n",
    ")\n",
    "\n",
    "ranked_genes_df.to_csv(\"/scratch/groups/icobos/Akila/markers/markers.csv\")    \n",
    "adata.write(\"/scratch/groups/icobos/Akila/markers/integrated_harmony_adata.h5ad\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
