{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "42169dff",
   "metadata": {},
   "source": [
    "# This pynb files contains\n",
    "\n",
    "1. Linear mixed model (LMM)\n",
    "\n",
    "2. Bootstrap Analysis (50% cells)\n",
    "\n",
    "3. Pseudobulk Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70a0ab0e",
   "metadata": {},
   "source": [
    "# Linear mixed model for identifying differential expression genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06231c45",
   "metadata": {},
   "source": [
    "# comparison between low vs int (early) disease group from neocortex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73426bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load R packages\n",
    "library('Matrix')\n",
    "library('dplyr')\n",
    "library('ggplot2')\n",
    "library('MAST')\n",
    "library(\"lme4\",lib.loc=\"/home/users/akila20/R/x86_64-pc-linux-gnu-library/4.0/\")\n",
    "library('SingleCellExperiment')\n",
    "library('zellkonverter')\n",
    "# Write output \n",
    "OUT<-'/scratch/groups/icobos/Akila/recompute/DEG/BA9/unspliced/low_int/'\n",
    "\n",
    "#Read the h5ad file convert to SCE \n",
    "\n",
    "## anndata contains low,int disease group from BA9\n",
    "example_h5ad <- \"/scratch/groups/icobos/Akila/recompute/DEG/BA9/BA9_low_int.h5ad\"\n",
    "sc=readH5AD(example_h5ad)\n",
    "\n",
    "#### use the raw counts \n",
    "counts(sc) <- assay(sc, \"unspliced\")\n",
    "\n",
    "#iterate over each celltype and identfy DEG\n",
    "cluster<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\t\"Oligodendrocyte-OPALIN\",\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster1<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\"Oligodendrocyte-OPALIN\",\t\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster_count=length(cluster)\n",
    "\n",
    "## do the mixed model for all the clusters\n",
    "for (i in 1:cluster_count)\n",
    "{\n",
    "\n",
    "sce<-sc[, sc$Author_Annotation == cluster[i]]\n",
    "\n",
    "\n",
    "logcounts(sce) <- log2(counts(sce) + 1) #log2 transform the raw data**\n",
    "\n",
    "sca <- SceToSingleCellAssay(sce) # I then upcast the SingleCellExperiment object to MAST’s subclass SingleCellAssay\n",
    "\n",
    "# cellular detection rate\n",
    "cdr <-colSums(assay(sca)>0)\n",
    "colData(sca)$cngeneson <- scale(cdr) # I calculate CDR\n",
    "scaSample <- sca[sample(which(freq(sca)>.1), 20),]\n",
    "flat <- as(scaSample, 'data.table')\n",
    "thres <- thresholdSCRNACountMatrix(assay(sca),cutbins = NULL, nbins = 10, min_per_bin = 30,  bin_by = \"median\")\n",
    "assays(sca, withDimnames = FALSE) <- list(thresh=thres$counts_threshold, tpm=assay(sca))\n",
    "\n",
    "### set the minimum background\n",
    "freq_expressed <- 0.15\n",
    "expressed_genes <- freq(sca) > freq_expressed\n",
    "sca <- sca[expressed_genes,]\n",
    "\n",
    "#fetch the condition(LOW/INT)\n",
    "\n",
    "cond <-factor(colData(sca)$Disease.Group)\n",
    "\n",
    "#set the reference value for condition\n",
    "cond <-relevel(cond,'int') # In order to have more interpretable coefficients, I set the reference level of my genotype condition to be my wild type cells.\n",
    "colData(sca)$test <-cond\n",
    "print(sca@colData$test)\n",
    "\n",
    "\n",
    "saveRDS(sca,paste0(OUT,\"low_int\",str(cluster1[i]), \"_sca.rds\"))\n",
    "\n",
    "zlm_mixed <- zlm(~test + (1|(subject))  + cngeneson + Age + Sex + total_counts,sca, method = 'glm')\n",
    "saveRDS(zlm_mixed, paste0(OUT,\"low_int\",cluster1[i], \"_zlm_mixed.rds\"))\n",
    "\n",
    "#calculate logFC between all the fixed effect parameters\n",
    "logFC <- getLogFC(zlm_mixed)\n",
    "write.csv(logFC, paste0(OUT,\"low_int\",cluster1[i],\"_logFC_mixed.csv\"))\n",
    "\n",
    "#Log-likelihood model\n",
    "summary_mixed <- summary(zlm_mixed, doLRT= 'testlow') \n",
    "saveRDS(summary_mixed, paste0(OUT,\"low_int\",cluster1[i],\"_summarymixedLRT.rds\"))\n",
    "summarytest_mixed <- summary_mixed$datatable\n",
    "\n",
    "\n",
    "fcHurdletest_mixed <- merge(summarytest_mixed[contrast=='testlow' & component=='H',.(primerid, `Pr(>Chisq)`)], \n",
    "                      summarytest_mixed[contrast=='testlow' & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], \n",
    "                      by='primerid')\n",
    "\n",
    "                      \n",
    "fcHurdletest_mixed[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "\n",
    "write.csv(fcHurdletest_mixed, paste0(OUT,\"low_int\",cluster1[i],\"_mixed_Hurdle.csv\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe88fe7c",
   "metadata": {},
   "source": [
    "# comparison between int vs high disease group from neocortex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf8cc1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load R packages\n",
    "library('Matrix')\n",
    "library('dplyr')\n",
    "library('ggplot2')\n",
    "library('MAST')\n",
    "library(\"lme4\",lib=\"/home/users/akila20/R/x86_64-pc-linux-gnu-library/4.0/\")\n",
    "library('SingleCellExperiment')\n",
    "library('zellkonverter')\n",
    "# Write output \n",
    "\n",
    "OUT<-'/scratch/groups/icobos/Akila/recompute/DEG/BA9/unspliced/int_high/'\n",
    "\n",
    "## anndata contains int,high disease group from BA9\n",
    "example_h5ad <- \"/scratch/groups/icobos/Akila/recompute/DEG/BA9/BA9_int_high.h5ad\"\n",
    "sc=readH5AD(example_h5ad)\n",
    "counts(sc) <- assay(sc, \"unspliced\")\n",
    "\n",
    "#iterate over each celltype and identify DEG\n",
    "\n",
    "cluster<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\t\"Oligodendrocyte-OPALIN\",\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster1<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\"Oligodendrocyte-OPALIN\",\t\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster_count=length(cluster)\n",
    "\n",
    "## do the mixed model for all the clusters\n",
    "for (i in 1:cluster_count)\n",
    "{\n",
    "\n",
    "sce<-sc[, sc$Author_Annotation == cluster[i]]\n",
    "\n",
    "\n",
    "logcounts(sce) <- log2(counts(sce) + 1) #log2 transform the raw data**\n",
    "\n",
    "sca <- SceToSingleCellAssay(sce) # I then upcast the SingleCellExperiment object to MAST’s subclass SingleCellAssay\n",
    "\n",
    "# cellular detection rate\n",
    "cdr <-colSums(assay(sca)>0)\n",
    "colData(sca)$cngeneson <- scale(cdr) # I calculate CDR\n",
    "scaSample <- sca[sample(which(freq(sca)>.1), 20),]\n",
    "flat <- as(scaSample, 'data.table')\n",
    "thres <- thresholdSCRNACountMatrix(assay(sca),cutbins = NULL, nbins = 10, min_per_bin = 30,  bin_by = \"median\")\n",
    "assays(sca, withDimnames = FALSE) <- list(thresh=thres$counts_threshold, tpm=assay(sca))\n",
    "    \n",
    "### set the minimum background\n",
    "freq_expressed <- 0.15\n",
    "expressed_genes <- freq(sca) > freq_expressed\n",
    "sca <- sca[expressed_genes,]\n",
    "\n",
    "\n",
    "cond <-factor(colData(sca)$Disease.Group)\n",
    "\n",
    "#set the reference value for condition\n",
    "cond <-relevel(cond,'high') # In order to have more interpretable coefficients, I set the reference level of my genotype condition to be my wild type cells.\n",
    "colData(sca)$test <-cond\n",
    "print(sca@colData$test)\n",
    "\n",
    "\n",
    "saveRDS(sca,paste0(OUT,\"int_high\",cluster1[i], \"_sca.rds\"))\n",
    "\n",
    "zlm_mixed <- zlm(~test + (1|(subject)) + cngeneson  + Age + Sex  + total_counts,sca, method = 'glm')\n",
    "saveRDS(zlm_mixed, paste0(OUT,\"int_high\",cluster1[i], \"_zlm_mixed.rds\"))\n",
    "\n",
    "#calculate logFC between all the fixed effect parameters\n",
    "logFC <- getLogFC(zlm_mixed)\n",
    "write.csv(logFC, paste0(OUT,\"int_high\",cluster1[i],\"_logFC_mixed.csv\"))\n",
    "\n",
    "#Log-likelihood model\n",
    "summary_mixed <- summary(zlm_mixed, doLRT= 'testint') \n",
    "saveRDS(summary_mixed, paste0(OUT,\"int_high\",cluster1[i],\"_summarymixedLRT.rds\"))\n",
    "summarytest_mixed <- summary_mixed$datatable\n",
    "\n",
    "\n",
    "fcHurdletest_mixed <- merge(summarytest_mixed[contrast=='testint' & component=='H',.(primerid, `Pr(>Chisq)`)], \n",
    "                      summarytest_mixed[contrast=='testint' & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], \n",
    "                      by='primerid')\n",
    "\n",
    "                      \n",
    "fcHurdletest_mixed[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "\n",
    "write.csv(fcHurdletest_mixed, paste0(OUT,\"int_high\",cluster1[i],\"_mixed_Hurdle.csv\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d88ab7a",
   "metadata": {},
   "source": [
    "# bootstrap LMM comparison between low vs int Disease group (Neocortex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "880948dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load R packages\n",
    "library('Matrix')\n",
    "library('dplyr')\n",
    "library('ggplot2')\n",
    "library('MAST')\n",
    "library(\"lme4\")\n",
    "library('SingleCellExperiment')\n",
    "\n",
    "### specify celltypes for bootstrap\n",
    "\n",
    "cluster<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\t\"Oligodendrocyte-OPALIN\",\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster1<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\"Oligodendrocyte-OPALIN\",\t\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "### specify the bootstrap iterations\n",
    "n_iter=100\n",
    "cluster_count=length(cluster)\n",
    "OUT<-\"/scratch/groups/icobos/Akila/recompute/bootstrap/unspliced/low_int/\"\n",
    "## do the mixed model for all the clusters\n",
    "for (j in 1:cluster_count)\n",
    "{\n",
    "#### read the cell type specific rds files which contains low,int,high disease group\n",
    "sc<-readRDS(paste0(\"/scratch/groups/icobos/Akila/recompute/bootstrap/unspliced/files/\",cluster1[j],\".rds\"))\n",
    "iter_output <- list()\n",
    "### bootstrap_iterations\n",
    "for (i in 1:100)\n",
    "{\n",
    "  sce<-sc\n",
    "  \n",
    "  ### calculate the 50% cells from low,int disease group\n",
    "  low <- sce[, sce$Disease.Group == \"low\"]\n",
    "  inter<-sce[, sce$Disease.Group == \"int\"]\n",
    "  ### number of cells in int/low\n",
    "  n_int=ncol(inter)\n",
    "  n_low =ncol(low)\n",
    "    \n",
    "  #### fetch 50% cells from low, int disease group for 100 iterations\n",
    "  frac_to_keep <- 0.5\n",
    "  keep_low <- floor(n_low*frac_to_keep)\n",
    "  keep_int <- floor(n_inter*frac_to_keep)\n",
    "  \n",
    "  sce_int <- inter[,sample(colnames(inter),keep_int)]\n",
    "  sce_low <- low[,sample(colnames(low),keep_low)]\n",
    "  \n",
    "  \n",
    "  sce_merged <- cbind(sce_int,sce_low)\n",
    "  logcounts(sce_merged) <- log2(counts(sce_merged) + 1) #log2 transform the raw data**\n",
    "  \n",
    "  sca <- SceToSingleCellAssay(sce_merged) # I then upcast the SingleCellExperiment object to MAST’s subclass SingleCellAssay\n",
    "  \n",
    "  # cellular detection rate\n",
    "  cdr <-colSums(assay(sca)>0)\n",
    "  colData(sca)$cngeneson <- scale(cdr) # I calculate CDR\n",
    "  scaSample <- sca[sample(which(freq(sca)>.1), 20),]\n",
    "  flat <- as(scaSample, 'data.table')\n",
    "  thres <- thresholdSCRNACountMatrix(assay(sca),cutbins = NULL, nbins = 10, min_per_bin = 30,  bin_by = \"median\")\n",
    "  assays(sca, withDimnames = FALSE) <- list(thresh=thres$counts_threshold, tpm=assay(sca))\n",
    "  freq_expressed <- 0.15\n",
    "  expressed_genes <- freq(sca) > freq_expressed\n",
    "  sca <- sca[expressed_genes,]\n",
    "  \n",
    "  #fetch the condition(LOW/INT)\n",
    "  \n",
    "  cond <-factor(colData(sca)$Disease.Group)\n",
    "  \n",
    "  #set the reference value for condition\n",
    "  cond <-relevel(cond,'int') # In order to have more interpretable coefficients, I set the reference level of my genotype condition to be my wild type cells.\n",
    "  colData(sca)$test <-cond\n",
    "  print(sca@colData$test)\n",
    "  zlm_mixed <- zlm(~test + (1|(subject)) + cngeneson + Age + Sex + total_counts,sca, method = 'glm')\n",
    "\n",
    "  #calculate logFC between all the fixed effect parameters\n",
    "  FC <-getLogFC(zlm_mixed)\n",
    "  filter_FC <- subset(FC, contrast == \"testlow\")\n",
    " \n",
    "  \n",
    "  #Log-likelihood model\n",
    "  summary_mixed <- summary(zlm_mixed, doLRT= 'testlow') \n",
    "  summarytest_mixed <- summary_mixed$datatable\n",
    "  \n",
    "  \n",
    "  fcHurdletest_mixed <- merge(summarytest_mixed[contrast=='testlow' & component=='H',.(primerid, `Pr(>Chisq)`)], \n",
    "                              summarytest_mixed[contrast=='testlow' & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], \n",
    "                              by='primerid')\n",
    "  \n",
    "  \n",
    "  de<-fcHurdletest_mixed[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "  \n",
    "  de$logFC <- filter_FC[filter_FC$primerid %in% de$primerid,3]\n",
    "  iter_output[[i]] <- de\n",
    "  print(i)\n",
    "}\n",
    "### filter FDR <0.05\n",
    "\n",
    "result <- lapply(iter_output,function(x) x[x$fdr < 0.05,] )\n",
    "\n",
    "###create matrix for logFC and FDR for the cluster\n",
    "\n",
    "list_of_genes <- unique(do.call(c,lapply(result,function(x) data.frame(x[[1]])[,1])))\n",
    "result_matrix_log_fold <- data.frame(matrix(NA,nrow = length(list_of_genes),ncol = n_iter),row.names = list_of_genes)\n",
    "result_matrix_log_q <- data.frame(matrix(NA,nrow = length(list_of_genes),ncol = n_iter),row.names = list_of_genes )\n",
    "\n",
    "for(k in 1:length(result)){\n",
    "  ### get the unique differentially expressed genes from all iterations\n",
    "  keys_gene <- data.frame(result[[k]])[,1]\n",
    "  #### fetch logFC and FDR value from each iterations to matrix\n",
    "  result_matrix_log_fold[keys_gene,k] <- data.frame(result[[k]])[,7]\n",
    "  result_matrix_log_q[keys_gene,k] <- data.frame(result[[k]])[,6]\n",
    "}\n",
    "### Find the mean value of logFC and FDR\n",
    "result_matrix_log_fold$mean_logFC <- unname(apply(result_matrix_log_fold,1,function(x) mean(x,na.rm = TRUE)))\n",
    "result_matrix_log_q$mean_qvalue <- unname(apply(result_matrix_log_q,1,function(x) mean(x,na.rm = TRUE)))\n",
    "\n",
    "write.csv(result_matrix_log_fold,paste0(OUT,\"low_int\",cluster1[j],\"_logFC_iteration.csv\"))\n",
    "write.csv(result_matrix_log_q,paste0(OUT,\"low_int\",cluster1[j],\"qvalue_iteration.csv\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72fc36f3",
   "metadata": {},
   "source": [
    "# bootstrap LMM comparison between int vs high Disease group (Neocortex)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29535918",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load R packages\n",
    "library('Matrix')\n",
    "library('dplyr')\n",
    "library('ggplot2')\n",
    "library('MAST')\n",
    "library(\"lme4\")\n",
    "library('SingleCellExperiment')\n",
    "\n",
    "\n",
    "### specify celltypes for bootstrap\n",
    "\n",
    "cluster<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\t\"Oligodendrocyte-OPALIN\",\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster1<-c(\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-GFAP-VCAN\",\t\"OPC\",\t\"Microglia-Reactive-CD163\",\t\"Microglia-Homeostatic-CX3CR1\",\t\"Oligodendrocyte-OPALIN\",\t\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex6:[RORB-MME]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\")\n",
    "\n",
    "cluster_count=length(cluster)\n",
    "OUT<-\"/scratch/groups/icobos/Akila/recompute/bootstrap/unspliced/int_high/\"\n",
    "## do the mixed model for all the clusters\n",
    "for (j in 1:cluster_count)\n",
    "{\n",
    "#### read the cell type specific rds files which contains low,int,high disease group\n",
    "\n",
    "sc<-readRDS(paste0(\"/scratch/groups/icobos/Akila/recompute/bootstrap/unspliced/files/\",cluster1[j],\".rds\"))\n",
    "iter_output <- list()\n",
    "    \n",
    "### specify the bootstrap iterations\n",
    "for (i in 1:100)\n",
    "{\n",
    "  sce<-sc\n",
    "  \n",
    "######calculate the 50% cells from int, high disease group\n",
    "    \n",
    "  inter <- sce[, sce$Disease.Group == \"int\"]\n",
    "  high<-sce[, sce$Disease.Group == \"high\"]\n",
    " \n",
    "  ### number of cells in int/high\n",
    "    \n",
    "  n_int=ncol(inter)\n",
    "  n_high=ncol(high)\n",
    "    \n",
    "  #### fetch 50% cells from int, high disease group for 100 iterations\n",
    "\n",
    "  frac_to_keep <- 0.5\n",
    "  keep_int <- floor(n_int*frac_to_keep)\n",
    "  keep_high <- floor(n_high*frac_to_keep)\n",
    "  \n",
    "  sce_int <- inter[,sample(colnames(inter),keep_int)]\n",
    "  sce_high <- high[,sample(colnames(high),keep_high)]\n",
    "  \n",
    "  \n",
    "  sce_merged <- cbind(sce_int,sce_high)\n",
    "  logcounts(sce_merged) <- log2(counts(sce_merged) + 1) #log2 transform the raw data**\n",
    "  \n",
    "  sca <- SceToSingleCellAssay(sce_merged) # I then upcast the SingleCellExperiment object to MAST’s subclass SingleCellAssay\n",
    "  \n",
    "  # cellular detection rate\n",
    "  cdr <-colSums(assay(sca)>0)\n",
    "  colData(sca)$cngeneson <- scale(cdr) # I calculate CDR\n",
    "  scaSample <- sca[sample(which(freq(sca)>.1), 20),]\n",
    "  flat <- as(scaSample, 'data.table')\n",
    "  thres <- thresholdSCRNACountMatrix(assay(sca),cutbins = NULL, nbins = 10, min_per_bin = 30,  bin_by = \"median\")\n",
    "  assays(sca, withDimnames = FALSE) <- list(thresh=thres$counts_threshold, tpm=assay(sca))\n",
    "  freq_expressed <- 0.15\n",
    "  expressed_genes <- freq(sca) > freq_expressed\n",
    "  sca <- sca[expressed_genes,]\n",
    "  \n",
    "  #fetch the condition(INT/HIGH)\n",
    "  \n",
    "  cond <-factor(colData(sca)$Disease.Group)\n",
    "  \n",
    "  #set the reference value for condition\n",
    "  cond <-relevel(cond,'high') # In order to have more interpretable coefficients, I set the reference level of my genotype condition to be my wild type cells.\n",
    "  colData(sca)$test <-cond\n",
    "  \n",
    "  zlm_mixed <- zlm(~test + (1|(subject)) + cngeneson + Age + Sex + RIN + total_counts,sca, method = 'glm')\n",
    " \n",
    "\n",
    "  #calculate logFC between all the fixed effect parameters\n",
    "  FC <-getLogFC(zlm_mixed)\n",
    "  filter_FC <- subset(FC, contrast == \"testint\")\n",
    "    \n",
    "   \n",
    "  #Log-likelihood model\n",
    "  summary_mixed <- summary(zlm_mixed, doLRT= 'testint') \n",
    "  summarytest_mixed <- summary_mixed$datatable\n",
    "  \n",
    "  \n",
    "  fcHurdletest_mixed <- merge(summarytest_mixed[contrast=='testint' & component=='H',.(primerid, `Pr(>Chisq)`)], \n",
    "                              summarytest_mixed[contrast=='testint' & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], \n",
    "                              by='primerid')\n",
    "  \n",
    "  \n",
    "  de<-fcHurdletest_mixed[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "  \n",
    "    de$logFC <- filter_FC[filter_FC$primerid %in% de$primerid,3]\n",
    "  iter_output[[i]] <- de\n",
    "  print(i)\n",
    "  \n",
    "}\n",
    "### filter FDR <0.05\n",
    "\n",
    "result <- lapply(iter_output,function(x) x[x$fdr < 0.05,] )\n",
    "\n",
    "###create matrix for logFC and FDR for the cluster\n",
    "n_iter=100\n",
    "list_of_genes <- unique(do.call(c,lapply(result,function(x) data.frame(x[[1]])[,1])))\n",
    "result_matrix_log_fold <- data.frame(matrix(NA,nrow = length(list_of_genes),ncol = n_iter),row.names = list_of_genes)\n",
    "result_matrix_log_q <- data.frame(matrix(NA,nrow = length(list_of_genes),ncol = n_iter),row.names = list_of_genes )\n",
    "print(cluster1[j])\n",
    "for(k in 1:length(result)){\n",
    "  ### get the unique differentially expressed genes from all iterations\n",
    "  keys_gene <- data.frame(result[[k]])[,1]\n",
    "  #### fetch logFC and FDR value from each iterations to matrix\n",
    "  result_matrix_log_fold[keys_gene,k] <- data.frame(result[[k]])[,7]\n",
    "  result_matrix_log_q[keys_gene,k] <- data.frame(result[[k]])[,6]\n",
    "}\n",
    "### Find the mean value of logFC and FDR\n",
    "result_matrix_log_fold$mean_logFC <- unname(apply(result_matrix_log_fold,1,function(x) mean(x,na.rm = TRUE)))\n",
    "result_matrix_log_q$mean_qvalue <- unname(apply(result_matrix_log_q,1,function(x) mean(x,na.rm = TRUE)))\n",
    "\n",
    "write.csv(result_matrix_log_fold,paste0(OUT,\"int_high\",cluster1[j],\"_logFC_iteration.csv\"))\n",
    "write.csv(result_matrix_log_q,paste0(OUT,\"int_high\",cluster1[j],\"qvalue_iteration.csv\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c903af1f",
   "metadata": {},
   "source": [
    "# pseudobulk comparison between (low vs Int-> Disease Group) (Neocortex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e996a15c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "import pydeseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "##load data\n",
    "bdata=sc.read_h5ad(\"/scratch/groups/icobos/Akila/recompute/QC_test/dataset.h5ad\")\n",
    "\n",
    "#### selet BA9\n",
    "bdata=bdata[bdata.obs[\"Brain.Region\"].isin([\"Frontal Cx (BA9)\"])]\n",
    "\n",
    "##select Early pathology\n",
    "early=bdata[bdata.obs[\"Disease.Group\"].isin([\"low\",\"int\"])]\n",
    "\n",
    "##define celltypes\n",
    "list1=[\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-GFAP-VCAN\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex6:[RORB-MME]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"Microglia-Homeostatic-CX3CR1\",\t\"Microglia-Reactive-AIF1\",\t\"Microglia-Reactive-CACNA1B\",\t\"Microglia-Reactive-CD163\",\t\"OPC\",\t\"Oligodendrocyte-COL18A1\",\t\"Oligodendrocyte-OPALIN\",\t\"Pericytes\",\t\"VLMC\",\t\"endothelial\",]\n",
    "\n",
    "for i in range(len(list1)):\n",
    "    ##subset each celltype\n",
    "    cell_subset=early[early.obs[\"Author_Annotation\"].isin([list1[i]])]\n",
    "    ##create peudobulk aggregate based on each subject\n",
    "    pbs=[]\n",
    "    for sample in cell_subset.obs[\"subject\"].unique():\n",
    "        samp_cell_subset=cell_subset[cell_subset.obs[\"subject\"] == sample]\n",
    "        samp_cell_subset.X=samp_cell_subset.layers[\"unspliced\"]\n",
    "        rep_adata =sc.AnnData(X=samp_cell_subset.X.sum(axis=0),var=samp_cell_subset.var[[]])\n",
    "        rep_adata.obs_names=[sample]\n",
    "        rep_adata.obs[\"Disease.Group\"]=samp_cell_subset.obs[\"Disease.Group\"].iloc[0]\n",
    "        pbs.append(rep_adata)\n",
    "        \n",
    "    ## create counts for DDS\n",
    "    pb=sc.concat(pbs)\n",
    "    import pydeseq2\n",
    "    from pydeseq2.utils import wald_test\n",
    "    from pydeseq2.dds import DeseqDataSet\n",
    "    from pydeseq2.ds import DeseqStats\n",
    "    counts=pd.DataFrame(pb.X,columns=pb.var_names)\n",
    "    dds=DeseqDataSet(counts=counts,clinical=pb.obs,design_factors= \"Disease.Group\")\n",
    "    sc.pp.normalize_total(dds)\n",
    "    sc.pp.log1p(dds)\n",
    "    sc.pp.filter_genes(dds,min_cells=10)\n",
    "    from pydeseq2.utils import test_valid_counts\n",
    "    test_valid_counts(counts)\n",
    "    dds=DeseqDataSet(counts=counts,clinical=pb.obs,design_factors= \"Disease.Group\")\n",
    "    sc.pp.filter_genes(dds,min_cells=10)\n",
    "    dds.deseq2()\n",
    "    sc.tl.pca(dds)\n",
    "    stat_res = DeseqStats(dds, contrast=(\"Disease.Group\", \"int\", \"low\"),alpha=0.05)\n",
    "    stat_res.summary()\n",
    "    de = stat_res.results_df\n",
    "    de.sort_values('stat', ascending=False)\n",
    "    de.to_csv(\"early/unspliced/BA9/\"+str(list1[i])+\"_pseudobulk.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54063a4f",
   "metadata": {},
   "source": [
    "# pseudobulk comparison between (Int vs High-> Disease Group) (Neocortex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "227063c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "import pydeseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "##load data\n",
    "bdata=sc.read_h5ad(\"/scratch/groups/icobos/Akila/recompute/QC_test/dataset.h5ad\")\n",
    "\n",
    "#### selet BA9\n",
    "bdata=bdata[bdata.obs[\"Brain.Region\"].isin([\"Frontal Cx (BA9)\"])]\n",
    "\n",
    "#### select disease group\n",
    "late=bdata[bdata.obs[\"Disease.Group\"].isin([\"int\",\"high\"])]\n",
    "##define celltypes\n",
    "list1=[\"Astrocyte-GFAP-OSMR\",\t\"Astrocyte-GFAP-VCAN\",\t\"Astrocyte-SLC1A2-SMTN\",\t\"Astrocyte-SLC1A2-WIF1\",\t\"Ex1:[CUX2-SERPINE2-LAMP5]\",\t\"Ex2:[CUX2-RORB-GLIS3-LRRC2]\",\t\"Ex3:[CUX2-RORB-GLIS3-SV2C]\",\t\"Ex4:[CUX2-RORB-COL5A2-CLMN]\",\t\"Ex5:[RORB-CUX2-EYA4]\",\t\"Ex6:[RORB-MME]\",\t\"Ex7:[RORB-GABRG1]\",\t\"Ex8:[RORB-TLL1-TMTC4]\",\t\"Ex9:[RORB-TLL1-PCP4-MID1]\",\t\"Ex10:[RORB-TLL1-PCP4-TYR]\",\t\"Ex11:[RORB-ADGRL4]\",\t\"Ex12:[THEMIS-PRRX1]\",\t\"Ex13:[THEMIS-POSTN]\",\t\"Ex14:|FEZF2-HTR2C]\",\t\"Ex15:[FEZF2-ADRA1A]\",\t\"Ex16:[FEZF2-SYT6]\",\t\"Ex17:[FEZF2-ZFHX3-SEMA3D]\",\t\"Ex18:[FEZF2-ZFHX3-SCUBE1]\",\t\"In1:[LHX6-PVALB-CLMN]\",\t\"In2:[LHX6-PVALB-MEPE]\",\t\"In3:[LHX6-PVALB-COL15A1]\",\t\"In4:[LHX6-SST-MSR1]\",\t\"In5:[LHX6-SST-SPON1]\",\t\"In6:[LHX6-SST-SLC9A2]\",\t\"In7:[LHX6-SST-B3GAT2]\",\t\"In8:[LHX6-SST-SEL1L3]\",\t\"In9:[LHX6-SST-NPY]\",\t\"In10:[LHX6-ADARB2-LAMP5-HCRTR2]\",\t\"In11:[ADARB2-LAMP5-KIT]\",\t\"In12:[ADARB2-LAMP5-NDNF]\",\t\"In13:[ADARB2-SYT6]\",\t\"In14:[ADARB2-SEMA3C]\",\t\"In15:[ADARB2-VIP-DACH2]\",\t\"In16:[ADARB2-VIP-TAC3-NRP1]\",\t\"In17:[ADARB2-VIP-ABI3BP]\",\t\"In18:[ADARB2-VIP-EXPH5]\",\t\"In19:[ADARB2-CALB2-SCML4]\",\t\"Microglia-Homeostatic-CX3CR1\",\t\"Microglia-Reactive-AIF1\",\t\"Microglia-Reactive-CACNA1B\",\t\"Microglia-Reactive-CD163\",\t\"OPC\",\t\"Oligodendrocyte-COL18A1\",\t\"Oligodendrocyte-OPALIN\",\t\"Pericytes\",\t\"VLMC\",\t\"endothelial\",]\n",
    "\n",
    "for i in range(len(list1)):\n",
    "    ##subset each celltype\n",
    "    cell_subset=late[late.obs[\"author_annotation\"].isin([list1[i]])]\n",
    "    ##create peudobulk aggregate based on each subject\n",
    "    pbs=[]\n",
    "    for sample in cell_subset.obs[\"subject\"].unique():\n",
    "        samp_cell_subset=cell_subset[cell_subset.obs[\"subject\"] == sample]\n",
    "        samp_cell_subset.X=samp_cell_subset.layers[\"unspliced\"]\n",
    "        rep_adata =sc.AnnData(X=samp_cell_subset.X.sum(axis=0),var=samp_cell_subset.var[[]])\n",
    "        rep_adata.obs_names=[sample]\n",
    "        rep_adata.obs[\"Disease.Group\"]=samp_cell_subset.obs[\"Disease.Group\"].iloc[0]\n",
    "        pbs.append(rep_adata)\n",
    "    ## create counts for DDS\n",
    "    pb=sc.concat(pbs)\n",
    "    import pydeseq2\n",
    "    from pydeseq2.utils import wald_test\n",
    "    from pydeseq2.dds import DeseqDataSet\n",
    "    from pydeseq2.ds import DeseqStats\n",
    "    counts=pd.DataFrame(pb.X,columns=pb.var_names)\n",
    "    dds=DeseqDataSet(counts=counts,clinical=pb.obs,design_factors= \"Disease.Group\")\n",
    "    sc.pp.normalize_total(dds)\n",
    "    sc.pp.log1p(dds)\n",
    "    sc.pp.filter_genes(dds,min_cells=10)\n",
    "    from pydeseq2.utils import test_valid_counts\n",
    "    test_valid_counts(counts)\n",
    "    dds=DeseqDataSet(counts=counts,clinical=pb.obs,design_factors= \"Disease.Group\")\n",
    "    sc.pp.filter_genes(dds,min_cells=10)\n",
    "    dds.deseq2()\n",
    "    sc.tl.pca(dds)\n",
    "    stat_res = DeseqStats(dds, contrast=(\"Disease.Group\", \"high\", \"int\"),alpha=0.05)\n",
    "    stat_res.summary()\n",
    "    de = stat_res.results_df\n",
    "    de.sort_values('stat', ascending=False)\n",
    "    de.to_csv(\"late/unspliced/BA9/\"+str(list1[i])+\"_pseudobulk.csv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
